Extends PreparedTriggers.

oblitum opened the pull request:
Allows one to obtain the current trigger regex from inside semantic
completers, enabling customized completions based on it, like normal
code completion or argument hints.

<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/valloric/ycmd/305)
<!-- Reviewable:end -->

puremourning commented:
LGTM in isolation, though to me it seems strange to merge without seeing the usage of it.

Do you have a gist or something where you're using this in your fork? Or a quick explanation of how the completer uses it in a concrete case ?

Thanks,
Ben

---

Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/305#-:-K7krZEALI5OWW4QZfnC:332803040)*
<!-- Sent from Reviewable.io -->

oblitum commented:
@puremourning this is my current usage of it in the python part of the swift completer: https://gist.github.com/oblitum/8995c3cca765f06b3db1.

This change still requires a last minor change to ycmd, which I'll propose afterwards, to be entirely usable.

---

Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/305#-:-K7kzirQC9sJjV9Pzw78:-587492937)*
<!-- Sent from Reviewable.io -->

oblitum commented:
The YCM side also needs its own changes to have hints fully working, even though the server part gets done, the client is currently limiting itself in some points.

---

Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/305#-:-K7l1gECvZvimmHh4VLi:474543482)*
<!-- Sent from Reviewable.io -->

oblitum commented:
Also, with this change I'm not advocating this should be the final and only way to have multiplexed triggers, it's just the change the provides the least amount of effort, that's minimal, that doesn't interfere with the rest of the codebase and still allows completers to make use of it, if they care, in isolation. The user should get the option to have or not have different kinds of triggers just by adding/removing triggers from the triggers list, with the completer itself with the responsability of interpreting triggers. 

---

Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/305#-:-K7l4MYKT6ov4wGUd8XT:-122779062)*
<!-- Sent from Reviewable.io -->

micbou commented:
I added minor comments. Looks good.

---

Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[ycmd/completers/completer_utils.py, line 48 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/305#-K7nMRUgdGhF0sIYCpZ5:-K7nMRUgdGhF0sIYCpZ6:919749606) ([raw file](https://github.com/valloric/ycmd/blob/4d2a61f4728a00dc464563d7fd43a4627c29a10d/ycmd/completers/completer_utils.py#L48)):*
Line above 80 characters.

---

*[ycmd/completers/completer_utils.py, line 100 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/305#-K7nMU8SkHWCuL2e_8Eh:-K7nMU8Tj03oqfyKv-C_:919749606) ([raw file](https://github.com/valloric/ycmd/blob/4d2a61f4728a00dc464563d7fd43a4627c29a10d/ycmd/completers/completer_utils.py#L100)):*
Line above 80 characters.

---

*[ycmd/completers/completer_utils_test.py, line 128 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/305#-K7nMcCWbJELz69zDtWt:-K7nMcCWbJELz69zDtWu:1205552486) ([raw file](https://github.com/valloric/ycmd/blob/4d2a61f4728a00dc464563d7fd43a4627c29a10d/ycmd/completers/completer_utils_test.py#L128)):*
Spaces after and before square brackets.

---

*[ycmd/completers/completer_utils_test.py, line 129 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/305#-K7nM_zconWeUSnp9CMD:-K7nM_zd8MXQN0HA7fYt:-556787841) ([raw file](https://github.com/valloric/ycmd/blob/4d2a61f4728a00dc464563d7fd43a4627c29a10d/ycmd/completers/completer_utils_test.py#L129)):*
Missing space before `)`.

---

*[ycmd/completers/completer_utils_test.py, line 131 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/305#-K7nMkvnIGu-UZqW7PST:-K7nMkvpiRK37dGbXxoB:-556787841) ([raw file](https://github.com/valloric/ycmd/blob/4d2a61f4728a00dc464563d7fd43a4627c29a10d/ycmd/completers/completer_utils_test.py#L131)):*
Missing space before `)`.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/305#-:-K7nNVEDN3agAsn-tTfP:912967832)*
<!-- Sent from Reviewable.io -->

oblitum commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[ycmd/completers/completer_utils.py, line 48 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/305#-K7nMRUgdGhF0sIYCpZ5:-K7nOrlDacHNd0EbnrB5:1754679759) ([raw file](https://github.com/valloric/ycmd/blob/4d2a61f4728a00dc464563d7fd43a4627c29a10d/ycmd/completers/completer_utils.py#L48)):*
ah, I thought I'd pass on this one... I'm not much used to python, dunno how to break this line without a backslash or breaking it into more statements =/ which I would not enjoy.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/305)*
<!-- Sent from Reviewable.io -->

oblitum commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[ycmd/completers/completer_utils.py, line 48 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/305#-K7nMRUgdGhF0sIYCpZ5:-K7nPrxlblKLks_NFK_J:429492428) ([raw file](https://github.com/valloric/ycmd/blob/4d2a61f4728a00dc464563d7fd43a4627c29a10d/ycmd/completers/completer_utils.py#L48)):*
ah, I guess I may break it at the arg list.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/305)*
<!-- Sent from Reviewable.io -->

oblitum commented:
Review status: 1 of 2 files reviewed at latest revision, 5 unresolved discussions.

---

*[ycmd/completers/completer_utils.py, line 48 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/305#-K7nMRUgdGhF0sIYCpZ5:-K7nRLv0GHx86XSmOhVA:-1620474753) ([raw file](https://github.com/valloric/ycmd/blob/4d2a61f4728a00dc464563d7fd43a4627c29a10d/ycmd/completers/completer_utils.py#L48)):*
fixed.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/305)*
<!-- Sent from Reviewable.io -->

oblitum commented:
Review status: 1 of 2 files reviewed at latest revision, 5 unresolved discussions.

---

*[ycmd/completers/completer_utils.py, line 100 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/305#-K7nMU8SkHWCuL2e_8Eh:-K7nRY30TIonCzkqvwbH:-1620474753) ([raw file](https://github.com/valloric/ycmd/blob/4d2a61f4728a00dc464563d7fd43a4627c29a10d/ycmd/completers/completer_utils.py#L100)):*
fixed.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/305)*
<!-- Sent from Reviewable.io -->

oblitum commented:
Review status: 0 of 2 files reviewed at latest revision, 5 unresolved discussions.

---

*[ycmd/completers/completer_utils_test.py, line 128 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/305#-K7nMcCWbJELz69zDtWt:-K7nSIh-6UpBUZmwGB7n:-1620474753) ([raw file](https://github.com/valloric/ycmd/blob/4d2a61f4728a00dc464563d7fd43a4627c29a10d/ycmd/completers/completer_utils_test.py#L128)):*
fixed.

---

*[ycmd/completers/completer_utils_test.py, line 129 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/305#-K7nM_zconWeUSnp9CMD:-K7nSLLenj4hw8BLaJmA:-1620474753) ([raw file](https://github.com/valloric/ycmd/blob/4d2a61f4728a00dc464563d7fd43a4627c29a10d/ycmd/completers/completer_utils_test.py#L129)):*
fixed.

---

*[ycmd/completers/completer_utils_test.py, line 131 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/305#-K7nMkvnIGu-UZqW7PST:-K7nSMcN1Y6Lvm_ppqvB:-1620474753) ([raw file](https://github.com/valloric/ycmd/blob/4d2a61f4728a00dc464563d7fd43a4627c29a10d/ycmd/completers/completer_utils_test.py#L131)):*
fixed.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/305)*
<!-- Sent from Reviewable.io -->

oblitum commented:
Thanks.

---

Review status: 0 of 2 files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/305#-:-K7nSQtXW8kd6Fy3EP_x:-1834867264)*
<!-- Sent from Reviewable.io -->

micbou commented:
Review status: 0 of 2 files reviewed at latest revision, 5 unresolved discussions.

---

*[ycmd/completers/completer_utils.py, line 48 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/305#-K7nMRUgdGhF0sIYCpZ5:-K7nQcOOs5seqT68KFGv:1413738837) ([raw file](https://github.com/valloric/ycmd/blob/4d2a61f4728a00dc464563d7fd43a4627c29a10d/ycmd/completers/completer_utils.py#L48)):*
You can also use parentheses. For example :
```python
return ( self.MatchingTriggerForFiletype( current_line, start_column, filetype )
              is not None )
```

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/305)*
<!-- Sent from Reviewable.io -->

oblitum commented:
Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion.

---

*[ycmd/completers/completer_utils.py, line 48 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/305#-K7nMRUgdGhF0sIYCpZ5:-K7nTWjSdox6dvPF2NNh:-563691857) ([raw file](https://github.com/valloric/ycmd/blob/4d2a61f4728a00dc464563d7fd43a4627c29a10d/ycmd/completers/completer_utils.py#L48)):*
ah yeah, I've recalled this now, I've seen this used to try dealing with long lambdas, which is sad by the way.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/305)*
<!-- Sent from Reviewable.io -->

Valloric commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

---

Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/305#-:-K7ngjVKZAi2ub9x23Cl:980077423)*
<!-- Sent from Reviewable.io -->

oblitum commented:
Anyone for the r+? :-)

---

Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/305#-:-K7rl565JH60o_uVO16C:800394325)*
<!-- Sent from Reviewable.io -->

vheon commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> @homu r+

---

Reviewed 1 of 1 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/305#-:-K7rlN-6j_pJ4lTsfLGg:-141632775)*
<!-- Sent from Reviewable.io -->

homu commented:
:pushpin: Commit 10802d8 has been approved by `vheon`

<!-- @homu r=vheon 10802d84932331713c95a11dcb1bea4442788f41 -->

homu commented:
:zap: Test exempted - [status](https://travis-ci.org/Valloric/ycmd/builds/101733270)

oblitum commented:
I've simplified the swift completer sample and it doesn't require the last ycmd change I commented about before. So that's it. That's all I'm using on ycmd side for hints.

---







*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/305#-:-K7s1AhAoyDcZrPAA5dM:-100988338)*
<!-- Sent from Reviewable.io -->

vheon commented:
You mean that this is the last change needed for the Swift completer including the argument hints feature?

oblitum commented:
Yes, on ycmd side, there're still changes on YCM side, some minor and fundamental, other not so minor but not fundamental. I'll proceed with them, you may take a look at the gist I've posted before.

---







*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/305#-:-K7s2eS2RwQ3-sjIVLkD:-1630964854)*
<!-- Sent from Reviewable.io -->
